### DISCUSSION TITLE
Confusion on which part of code is being executed on repeated job

### DISCUSSION THREAD

**Discussion by elhananjair** [taskforcesh/bullmq#2500] [Q&A]
Hello friends I am having some logical issues here on my code, my application repeatedly sends SMS to client each day at some scheduled time. But on the next call of the function inside the worker, it first shows that the job is completed (before it get started) and then some functions get executed before the others. Here is the code I am working on, 

```
     async function getScheduledTime() {
        return await sql`select scheduletime from setting.smsschedule`
    }
    
    getScheduledTime().then(val => {
        // I check the time the application starts before proceeding to start the sending sms function so that it won't be missed for that day incase the application starts after the scheduled time (Please correct me if this is not how bullmq works)
    });
    
      async function addJobs() {
        await send_sms_queue.add('sendSMS', {}, {
            repeat: {
                pattern: `${arrScheduledTime[2]} ${arrScheduledTime[1]} ${arrScheduledTime[0]} * * *`,
            }
        });
    }
    
```

Now I want to process cleaning the database before the main sending function starts but that's not happening.

```
const worker = new Worker('send_sms_queue', async () => {
           cleanupdb().then(async () => {
               if (!(weekly_name[today.getDay()] === 'Sunday')) {
                   await connectSMSC(); //This is the function I want to execute each day but only after cleaning the database
                   connectReceiver()
               } else if (weekly_name[today.getDay()] === 'Sunday') {
                   console.log(`Sending SMS in ${weekly_name[today.getDay()]} is forbidden!`)
               }
           })
       }, {
           connection: {
               host: 'localhost',
               port: 6379
           }
       });
       
       
    worker.on('completed', job => { //implement something like when all sms have sent resolve it so that it will be marked as resolved
           console.log(`${job.name} is completed!`) //I always see this output first 
       })
```

The first thing being shown on each day at the scheduled time is, `console.log(`${job.name} is completed!`)` from on complete event handler, and then it goes to `connectSMSC` function which should be called after `cleanupdb` function, as a result I am not getting what I was expecting on the flow of the application. 

I appreciate your support guys.

**Comment by manast** [comment]
It is difficult to know what is going on as you post several moving parts. I recommend you write some simpler testable code and then build up from there.

**Comment by elhananjair** [comment]
@manast ok sorry I didn't describe it well.

I am developing SMPP related application using sveltekit and I am using bullmq to execute `sending SMS` function in background. I have posted on another thread, https://github.com/taskforcesh/bullmq/discussions/2469

I have implemented the above code inside `hooks.server.js`
Here is the general idea and steps

1. Check schedule time in database when the application starts. (I did this in case the application starts after the schedule time based on bullmq repeatable job parameter. )
    1.1 If the current time is greater than the schedule time, the cleanupdatabse function get executed and then sending sms function will be executed. At the end of the block I executed adding a job to queue
    
2. If the application up and running without interruption, the sending function get executed after databaseclean up finished. 


That's is the whole idea. Here is the full implemented code

`hooks.server.js`

```
import { connectSMSC } from '$lib/server/sendsms.js'
import { connectReceiver} from "$lib/server/delivery.js";
import { cleanupdb } from '$lib/server/dbcleanup.js'
import { building } from '$app/environment';
import { Queue, Worker } from 'bullmq';
        
   getScheduledTime().then(val => {
             if (currentTime > scheduledtime) {
                            cleanupdb().then(async () => { 
                            await connectSMSC(); //Sends SMS 
                  })
              }
       addJobs().then(() => console.log("Jobs added successfully")); 
    });
    
      async function addJobs() {
        await send_sms_queue.add('sendSMS', {}, {
            repeat: {
                pattern: `${arrScheduledTime[2]} ${arrScheduledTime[1]} ${arrScheduledTime[0]} * * *`,
            }
        });
    }


    const send_sms_queue = new Queue('send_sms_queue', {
        connection: {
            host: 'localhost',
            port: 6379
        }
    });


       const worker = new Worker('send_sms_queue', async () => {
           cleanupdb().then(async () => {
               if (!(weekly_name[today.getDay()] === 'Sunday')) {
                   await connectSMSC();
                   connectReceiver()
               } else if (weekly_name[today.getDay()] === 'Sunday') {
                   console.log(`Sending SMS in ${weekly_name[today.getDay()]} is forbidden!`)
               }
           })
       }, {
           connection: {
               host: 'localhost',
               port: 6379
           }
       });

       worker.on('completed', job => {  
           console.log(`${job.name} is completed!`)
       })
    

```

**Reply by manast** [reply]
It is going to be difficult to debug this code here but out of the bat I am seeing that you are not correctly returning the promise of the call to ```cleanupdb``` so you loose any potential errors throw by it and also it is marking the job as completed even though it has not completed. This would be a better way to write it:
```ts
const worker = new Worker('send_sms_queue', async () => {
  await cleanupdb()
  if (!(weekly_name[today.getDay()] === 'Sunday')) {
    await connectSMSC();
    connectReceiver()
  } else if (weekly_name[today.getDay()] === 'Sunday') {
    console.log(`Sending SMS in ${weekly_name[today.getDay()]} is forbidden!`)
  }
}, {
  connection: {
    host: 'localhost',
    port: 6379
  }
});
```

Not sure if ```connectReceiver()``` is an async call too, in such case you should also await it.

**Reply by elhananjair** [reply]
@manast thank you so much for the reply, I will update my code according to your suggestion. 
But if you saw one function above `getScheduledTime`, I added that function because I wasn't sure if **bullmq** handles jobs that are haven't executed because of system issue or other reasons, therefore, I execute **connectSMSC** as the application get started. 

Let's say I set repeatable job to be executed always at 9:00 AM, and the application is not running at that moment and I manually started it at 9:15 AM. What does happen to the scheduled job? That's my question. That is the reason why I execute the `connectSMSC` function if the time the application started is greater than the scheduled time.

**Reply by elhananjair** [reply]
I think [this](https://github.com/OptimalBits/bull/issues/1239) issue might explain what I am trying to do with my application, I am trying to make the function immediately start instead of waiting for the scheduled time to be executed. I tried solving that by checking the time the application got executed and deciding whether to directly execute the function or wait for the job to be executed based on the scheduled time.

**Reply by manast** [reply]
Just try it, it will execute as soon as a worker is online since the job's time to be executed has passed.

**Reply by elhananjair** [reply]
I have tried it and only `Job added successfully` is printed out and it its from the following line and nothing after that
```
 addJobs().then(() => console.log("Jobs added successfully"));
```

**Reply by elhananjair** [reply]
![image](https://github.com/taskforcesh/bullmq/assets/69236638/301d3352-4e9d-40e9-8d2d-1ba8bc3234db)

**Reply by elhananjair** [reply]
based on this statement, it worker should be started as it gets instantiated, `When a worker instance is created, it launches the processor immediately` but that's not happening here.
